### PR TITLE
LWDEV-8299 Add context facility in SingleLimitOrderService

### PR DESCRIPTION
**Pull request by rocketmannot**

### CODE REVIEW

File: src/main/kotlin/com/lykke/matching/engine/services/validators/input/LimitOrderInputValidator.kt [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+interface LimitOrderInputValidator {
+    fun validateLimitOrder(singleLimitOrderParsedData: SingleLimitOrderParsedData)
+    fun validateStopOrder(singleLimitOrderParsedData: SingleLimitOrderParsedData)
+    fun validateLimitOrder(isTrustedClient: Boolean,
+                           order: LimitOrder,
+                           assetPair: AssetPair,
+                           baseAssetDisabled: Boolean,
+                           quotingAssetDisabled: Boolean,
+                           baseAsset: Asset)
+    fun checkMinVolume(order: Order, assetPair: AssetPair): Boolean
```
Comment (barsukovskii) [inline src/main/kotlin/com/lykke/matching/engine/services/validators/input/LimitOrderInputValidator.kt]: The same: looks like it is part of the previous validation methods of this interface. Also it used as utility method in other places, but i think it is not reason to place this method in this interface

File: src/test/kotlin/com/lykke/matching/engine/daos/OrderTest.kt [hunk 2]
```diff
@@ -22,48 +22,4 @@ class OrderTest {
                 BigDecimal.valueOf(0.1), BigDecimal.valueOf(0.2)))
         testDictionariesDatabaseAccessor.addAssetPair(AssetPair("BTCUSD", "BTC", "USD", 8))
     }
-
-    @Test
-    fun testCheckVolume() {
```
Comment (barsukovskii) [inline src/test/kotlin/com/lykke/matching/engine/daos/OrderTest.kt]: Remove this class - now it has no tests

File: src/main/kotlin/com/lykke/matching/engine/daos/LimitOrder.kt [hunk 3]
```diff
@@ -31,6 +31,29 @@ class LimitOrder(id: String,
...
+            limitOrder.lastMatchTime,
+            limitOrder.reservedLimitVolume,
+            limitOrder.fee as? LimitOrderFeeInstruction,
+            limitOrder.fees?.map { it as NewLimitOrderFeeInstruction },
+            limitOrder.type,
+            limitOrder.lowerLimitPrice,
+            limitOrder.lowerPrice,
+            limitOrder.upperLimitPrice,
+            limitOrder.upperPrice,
+            limitOrder.previousExternalId)
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/daos/LimitOrder.kt]: fixed

File: src/main/kotlin/com/lykke/matching/engine/daos/context/SingleLimitOrderContext.kt [hunk 4]
```diff
@@ -0,0 +1,88 @@
+package com.lykke.matching.engine.daos.context
+
+import com.lykke.matching.engine.daos.Asset
+import com.lykke.matching.engine.daos.AssetPair
+import com.lykke.matching.engine.daos.LimitOrder
+import com.lykke.matching.engine.deduplication.ProcessedMessage
+import com.lykke.matching.engine.services.validators.impl.OrderValidationResult
+import com.lykke.matching.engine.utils.NumberUtils
+
+class SingleLimitOrderContext(val uid: String?,
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/daos/context/SingleLimitOrderContext.kt]: fixed

File: src/main/kotlin/com/lykke/matching/engine/order/SingleLimitOrderProcessor.kt [hunk 5]
```diff
@@ -14,22 +13,19 @@ import com.lykke.matching.engine.services.GenericLimitOrderService
...
+    fun processLimitOrder(singleLimitContext: SingleLimitOrderContext,
                           now: Date,
-                          messageId: String,
-                          processedMessage: ProcessedMessage?,
                           payBackReserved: BigDecimal? = null,
                           messageWrapper: MessageWrapper? = null) {
-        val assetPair = assetsPairsHolder.getAssetPair(order.assetPairId)
+        val order = singleLimitContext.limitOrder
+        val assetPair = singleLimitContext.assetPair
         val limitAsset = if (order.isBuySide()) assetPair.quotingAssetId else assetPair.baseAssetId
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/order/SingleLimitOrderProcessor.kt]: fixed

File: src/main/kotlin/com/lykke/matching/engine/daos/context/SingleLimitOrderContext.kt [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+                              val messageId: String,
+                              var limitOrder: LimitOrder,
+                              val isCancelOrders: Boolean,
+                              val assetPair: AssetPair,
+                              val baseAsset: Asset,
+                              val quotingAsset: Asset,
+                              val baseAssetDisabled: Boolean,
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/daos/context/SingleLimitOrderContext.kt]: fixed

File: src/main/kotlin/com/lykke/matching/engine/order/StopLimitOrderProcessor.kt [hunk 7]
```diff
@@ -195,16 +145,79 @@ class StopLimitOrderProcessor(private val limitOrderService: GenericLimitOrderSe
...
+                                    orderValidationResult: OrderValidationResult,
+                                    balance: BigDecimal, reservedBalance: BigDecimal,
+                                    cancelVolume: BigDecimal,
+                                    ordersToCancel: List<LimitOrder>,
+                                    clientLimitOrdersReport: LimitOrdersReport,
+                                    now: Date) {
+        val limitOrder = singleLimitContext.limitOrder
+        val limitAsset = singleLimitContext.limitAsset
+
+        LOGGER.info("${orderInfo(limitOrder)} $orderValidationResult.message")
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/order/StopLimitOrderProcessor.kt]: fixed

File: src/main/kotlin/com/lykke/matching/engine/services/validators/business/LimitOrderBusinessValidator.kt [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+import com.lykke.matching.engine.services.AssetOrderBook
+import java.math.BigDecimal
+
+interface LimitOrderBusinessValidator {
+    fun performValidation(isTrustedClient: Boolean,
+                          order: LimitOrder,
+                          availableBalance: BigDecimal,
+                          limitVolume: BigDecimal,
+                          orderBook: AssetOrderBook)
+    fun validateBalance(availableBalance: BigDecimal, limitVolume: BigDecimal)
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/services/validators/business/LimitOrderBusinessValidator.kt]: `validateBalance` - moved to common utility class `OrderValidationUtils`

File: src/main/kotlin/com/lykke/matching/engine/services/validators/input/LimitOrderInputValidator.kt (hunk 1 above)
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/services/validators/input/LimitOrderInputValidator.kt]: moved to utility class

File: src/test/kotlin/com/lykke/matching/engine/daos/OrderTest.kt (hunk 2 above)
Comment (rocketmannot) [reply src/test/kotlin/com/lykke/matching/engine/daos/OrderTest.kt]: removed

File: src/main/kotlin/com/lykke/matching/engine/services/validators/input/impl/LimitOrderInputValidatorImpl.kt [hunk 9]
```diff
@@ -72,12 +68,6 @@ class LimitOrderInputValidatorImpl : LimitOrderInputValidator {
         }
     }
 
-    private fun validateAssets(baseAssetDisabled: Boolean, quotingAssetDisabled: Boolean) {
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/services/validators/input/impl/LimitOrderInputValidatorImpl.kt]: fixed

**Pull request merged**